---
title: "Taxonomy bacteria"
author: "Valentin Gfeller"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
#global knitter option
knitr::opts_chunk$set(fig.align = "center", echo = FALSE)

# load packages
library(tidyverse)
library(here)
library(phyloseq)
library(vegan)
library(readxl)
library(ggpubr)
library(geomtextpath)
library(ggbeeswarm)
library(knitr)
library(seqinr)
library(ape)
library(msa)
library(ggtree)
library(ggnewscale)

# functions were written for repeatedly used code (see "/Scripts/Functions.R")
# source functions
source(here("Scripts", "Functions.R"))


# set global option for ggplot2
theme_set(theme_bw(base_size = 12))
```

Investigation of the bacteria taxonomy at different ranks to characterize the root microbes present in our samples. 

# Import and check data
```{r include=FALSE}
load(here("Data", "Interim", "phy.bac_all.RData"))
load(here("Data", "Interim", "phy.bac.rel.RData"))
load(here("Data", "Interim", "phy.bac_red.RData"))

```



# Taxonomy
## Check taxanomy of OTUs
```{r}
# rename unassigned taxa from "NA" to unassigned
tax_table(phy.bac_red) <- tax_table(phy.bac_red) %>% data.frame() %>% 
  mutate(across(everything(), ~ case_when(is.na(.) ~ "unassigned", TRUE ~ .))) %>% 
  as.matrix()

# look at different taxonomic ranks
tax_table(phy.bac_red) %>% colnames()
tax_table(phy.bac_red) %>% .["bOTU1", ]
```

### Phylum
```{r}
phy.bac_all_phyl_ <- tax_glom(phy.bac_red, "Phylum")
phy.bac_all_phyl <- transform_sample_counts(phy.bac_all_phyl_, 
                                            function(x) 100 * x/sum(x)) 

# melt phyloseq object
phy.bac_all_phyl_melt <- psmelt(phy.bac_all_phyl) %>%
  mutate(Phylum = case_when(
         Abundance < 0.5 ~ "low_abundant",
         TRUE ~ Phylum))

phy.bac_all_phyl_melt$Phylum %>% unique()

# arrange
phy.bac_all_phyl_melt <- phy.bac_all_phyl_melt %>% arrange(genotype)
phy.bac_all_phyl_melt <- phy.bac_all_phyl_melt %>% arrange(emergence)


sort_phylum <- phy.bac_all_phyl_melt %>%
  group_by(Phylum) %>% 
  summarise(Abundance = mean(Abundance)) %>% 
  arrange(desc(Abundance)) %>% pull(Phylum)


  # plot grouped by genotype origin
p <- phy.bac_all_phyl_melt %>% 
  mutate(breeding_material = factor(breeding_material,
                                    levels = c("gb", "bm", "cv")),
         genotype = factor(genotype, 
                           levels = phy.bac_all_phyl_melt$genotype %>% unique()),
         Phylum = factor(Phylum, 
                         levels = sort_phylum)) %>%
  ggplot(aes(x = genotype, y = Abundance, fill = Phylum)) + 
  geom_bar(stat ="identity", alpha = 0.8) +
  xlab("") + 
  ylab("Relative abundance [%]") +
  facet_wrap(vars(breeding_material), scales = "free", nrow = 3,
             labeller = as_labeller(c(bm = "Breeding material",
                                      cv = "Cultivar",
                                      gb = "Gene bank"))) + 
  ggtitle("Bacteria") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1,
                                   size = 3),
        strip.background = element_rect(fill = "grey90")) +
  scale_fill_brewer(palette = "Dark2")
  
p


p <- p
saveRDS(p, here("Data", "Interim", "Figures", "Taxonomy_phyl_bacteria.rds"))

p 

# check over all relative abundance of individual phyla
phy.bac_all_phyl_melt %>%
  group_by(Phylum) %>% 
  summarise(Abundance = mean(Abundance)) %>% 
  arrange(desc(Abundance))

```


### Genus
```{r}
phy.bac_all_Genus_ <- tax_glom(phy.bac_red, "Genus") 
phy.bac_all_Genus <- transform_sample_counts(phy.bac_all_Genus_, function(x) 100 * x/sum(x)) 

# melt phyloseq object
phy.bac_all_Genus_melt <- psmelt(phy.bac_all_Genus) 

# arrange
phy.bac_all_Genus_melt <- phy.bac_all_Genus_melt %>% arrange(genotype)
phy.bac_all_Genus_melt <- phy.bac_all_Genus_melt %>% arrange(breeding_material)

# plot
p <- phy.bac_all_Genus_melt  %>% 
  mutate(Genus = case_when(
         Abundance < 5 ~ "low_abundant",
         TRUE ~ Genus)) %>%
  mutate(Genus = factor(Genus, levels = c("Achromobacter", "Acidovorax", 
                         "Allorhizobium-Neorhizobium-Pararhizobium-Rhizobium",
                         "Cellvibrio", "Dechloromonas", 
                         "Flavobacterium", "Massilia", "Pseudomonas", 
                         "Serratia", "Stenotrophomonas", "Streptomyces",
                         "unassigned", "low_abundant"))) %>%
  ggplot(aes(x = genotype, y = Abundance, fill = Genus)) + 
  geom_bar(stat ="identity", alpha = 0.8) +
  xlab("") + 
  ylab("Relative abundance [%]") +
  facet_wrap(vars(breeding_material), scales = "free", nrow = 3,
             labeller = as_labeller(c(bm = "Breeding material",
                                      cv = "Cultivar",
                                      gb = "Gene bank"))) + 
  ggtitle("Bacteria: Taxonomy")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1,
                                   size = 6))
p

# check relative abundance of individual genus
phy.bac_all_Genus_melt %>% 
  group_by(Genus, genotype) %>% 
  summarise(Abundance = sum(Abundance)) %>% 
  group_by(Genus) %>% 
  summarise(mean = mean(Abundance)) %>% 
  slice_max(mean, n = 10)


```

### Species
```{r}
phy.bac_all_Species_ <- tax_glom(phy.bac_red, "Species") 
phy.bac_all_Species <- transform_sample_counts(phy.bac_all_Species_, function(x) 100 * x/sum(x)) 

# melt phyloseq object
phy.bac_all_Species_melt <- psmelt(phy.bac_all_Species) 

# arrange
phy.bac_all_Species_melt <- phy.bac_all_Species_melt %>% arrange(genotype)
phy.bac_all_Species_melt <- phy.bac_all_Species_melt %>% arrange(breeding_material)

# plot
p <- phy.bac_all_Species_melt %>%  
  mutate(Species = case_when(
         Abundance < 1 ~ "low_abundant",
         TRUE ~ Species)) %>% 
  ggplot(aes(x = genotype, y = Abundance, fill = Species)) + 
  geom_bar(stat ="identity", alpha = 0.8) +
  xlab("") + 
  ylab("Relative abundance [%]") +
  facet_wrap(vars(breeding_material), scales = "free", nrow = 3,
             labeller = as_labeller(c(bm = "Breeding material",
                                      cv = "Cultivar",
                                      gb = "Gene bank"))) + 
  ggtitle("Bacteria: Taxonomy")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1,
                                   size = 6))
  
p

# check relative abundance of individual species
phy.bac_all_Species_melt %>%
  group_by(Species, genotype) %>% 
  summarise(Abundance = sum(Abundance)) %>% 
  group_by(Species) %>% 
  summarise(mean = mean(Abundance)) %>% 
  slice_max(mean, n = 15)

# check relative abundance of individual genus WITHOUT unassigned taxa
phy.bac_all_Species_ <- subset_taxa(phy.bac_all_Species_, 
                                    !Species %in% 
                                      c("unassigned", "uncultured_bacterium"))
phy.bac_all_Species <- transform_sample_counts(phy.bac_all_Species_, function(x) 100 * x/sum(x)) 
phy.bac_all_Species_melt <- psmelt(phy.bac_all_Species) 

phy.bac_all_Species_melt %>%
  group_by(Species, genotype) %>% 
  summarise(Abundance = sum(Abundance)) %>% 
  group_by(Species) %>% 
  summarise(mean = mean(Abundance)) %>% 
  slice_max(mean, n = 15)


```


```{r}

```

