---
title: "Combine composite figures"
author: "Valentin Gfeller"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
#global knitter option
knitr::opts_chunk$set(fig.align = "center", dev = 'svglite', echo = FALSE)

# load packages
library(tidyverse); packageVersion("tidyverse")
library(phyloseq); packageVersion("phyloseq")
library(vegan); packageVersion("vegan")
library(ape); packageVersion("ape")
library(here); packageVersion("here")
library(readxl); packageVersion("readxl")
library(ggpubr); packageVersion("ggpubr")
library(nlme); packageVersion("nlme")
library(lme4); packageVersion("lme4")
library(knitr); packageVersion("knitr")
library(cowplot); packageVersion("cowplot")
library(factoextra); packageVersion("factoextra")
library(FactoMineR); packageVersion("FactoMineR")
library(ggbeeswarm); packageVersion("ggbeeswarm")
library(car); packageVersion("car")
library(emmeans); packageVersion("emmeans")
library(cowplot); packageVersion("cowplot")
library(UpSetR); packageVersion("UpSetR")

# functions were written for repeatedly used code (see "/Scripts/Functions.R")
# source functions
source(here("Scripts", "Functions.R"))

# set global option for ggplot2
theme_set(theme_bw(base_size = 14))
```

# Combine Figures
## Figure 1
```{r}
p.fun <- readRDS(here("Data", "Interim", "Figures", "Shannon_both_fungi_stats.rds"))

p.bac <- readRDS(here("Data", "Interim", "Figures", "Shannon_both_bacteria_stats.rds"))

fig <- plot_grid(p.fun,
                 p.bac,
                 labels = NULL,
                 align = "v",
                 nrow = 2, rel_widths = c(1, 1))
fig

# save
png(here("Output", "Figure_1.png"), units = "mm", 
    width = 280, height = 260, res = 600)
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure 2
```{r}
# Load individual plots and legend
p.fun <- readRDS(here("Data", "Interim", "Figures", "PCoA_genotype_material_fungi_stat.rds"))
p.bac <- readRDS(here("Data", "Interim", "Figures", "PCoA_genotype_material_bacteria_stat.rds"))
legend <- readRDS(here("Data", "Interim", "Figures", "PCoA_genotype_material_fungi_legend.rds"))

# Combine plots
fig <- plot_grid(p.fun,
                 p.bac,
                 align = "h",
                 nrow = 1, rel_widths = c(1, 1))

fig

# add legend
fig <- plot_grid(fig, legend, nrow = 2, rel_heights = c(1, 0.05))

fig

# save
png(here("Output", "Figure_2.png"), units = "mm", 
    width = 183*1.9, height = 120, res = 600, bg = "white")
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure 3
```{r}
# Load individual plots and legend
p.a <- readRDS(here("Data", "Interim", "Figures", "PCoA_emerg_fungi.rds"))
p.b <- readRDS(here("Data", "Interim", "Figures", "PCoA_rri_bacteria.rds"))
p.c <- readRDS(here("Data", "Interim", "Figures", "PCoA_emerg_genotypes_fungi.rds"))
p.d <- readRDS(here("Data", "Interim", "Figures", "PCo_axis_vs_emergence_fungi.rds"))

# Combine plots
fig <- plot_grid(p.a,
                 p.b,
                 p.c,
                 p.d,
                 align = "h",
                 nrow = 2)

fig

# save
png(here("Output", "Figure_3.png"), units = "mm", 
    width = 183*2, height = 120*2, res = 600)
fig
dev.off()

rm(list = c("p.a", "p.b", "p.c", "p.d", "fig"))
```

## Figure 4
```{r}
# Load individual plot
p <- readRDS(here("Data", "Interim", "Figures", "Heritability_all_combined.rds"))

p
# save
png(here("Output", "Figure_4.png"), units = "mm", 
    width = 183, height = 150, res = 600)
p
dev.off()

rm(list = c("p"))
```

## Figure 5
```{r}
# Load individual plots and legend
p.a <- readRDS(here("Data", "Interim", "Figures", "UpsetR_fungi.rds"))
p.b <- readRDS(here("Data", "Interim", "Figures", "UpsetR_bacteria.rds"))
p.c <- readRDS(here("Data", "Interim", "Figures", "Heatmap_genotype_fungi.rds"))
p.d <- readRDS(here("Data", "Interim", "Figures", "Heatmap_genotype_bact.rds"))
p.e <- readRDS(here("Data", "Interim", "Figures", "Cor_OTU_emergence.rds"))

p.legend <- readRDS(here("Data", "Interim", "Figures", "Heatmap_genotype_fungi_legend.rds"))
p.legend_bac <- readRDS(here("Data", "Interim", "Figures", "Heatmap_genotype_bacteria_legend.rds"))


# Combine plots
fig <- plot_grid(p.a,
                 p.b,
                 p.c,
                 p.d,
                 align = "h", nrow = 2)

fig

# add legend
fig <- plot_grid(fig, plot_grid(p.legend, p.legend_bac, nrow = 1), p.e, nrow = 3, rel_heights = c(1, 0.075, 0.5))

fig

# save
png(here("Output", "Figure_5.png"), units = "mm", 
    width = 183*1.5, height = 120*3, res = 600)
fig
dev.off()

rm(list = c("p.a", "p.b",  "p.c", "p.d", "p.e", "fig"))
```

## Figure S02
```{r}
p.fun <- readRDS(here("Data", "Interim", "Figures", "Sequencing_depth_fungi.rds"))
p.bac <- readRDS(here("Data", "Interim", "Figures", "Sequencing_depth_bacteria.rds"))

fig <- plot_grid(p.fun,
                 p.bac,
                 labels = NULL,
                 align = "v",
                 nrow = 2, rel_widths = c(1, 1))
fig

# save
png(here("Output", "Figure_S02.png"), units = "mm", 
    width = 183, height = 155, res = 600)
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure S03
```{r}
p.fun <- readRDS(here("Data", "Interim", "Figures", "Taxonomy_phyl_fungi.rds"))
p.bac <- readRDS(here("Data", "Interim", "Figures", "Taxonomy_phyl_bacteria.rds"))

fig <- plot_grid(p.fun,
                 p.bac,
                 labels = NULL,
                 align = "v",
                 nrow = 2, rel_widths = c(1, 1))
fig

# save
png(here("Output", "Figure_S03.png"), units = "mm", 
    width = 183, height = 260, res = 600)
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure S04
```{r}
p.fun <- readRDS(here("Data", "Interim", "Figures", "Richness_both_fungi_stats.rds"))
p.bac <- readRDS(here("Data", "Interim", "Figures", "Richness_both_bacteria_stats.rds"))

fig <- plot_grid(p.fun,
                 p.bac,
                 labels = NULL,
                 align = "v",
                 nrow = 2, rel_widths = c(1, 1))
fig

# save
png(here("Output", "Figure_S04.png"), units = "mm", 
    width = 280, height = 260, res = 600)
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure S05
```{r}
# Load individual plots and legend
p.fun <- readRDS(here("Data", "Interim", "Figures", "Betadisper_fungi_stats.rds"))
p.bac <- readRDS(here("Data", "Interim", "Figures", "Betadisper_bacteria_stats.rds"))

# Combine plots
fig <- plot_grid(p.fun,
                 p.bac,
                 align = "h",
                 nrow = 2, rel_heights = c(1, 1))

fig

# save
# 89 mm wide (single column) and 183 mm wide (double column)
png(here("Output", "Figure_S05.png"), units = "mm", 
   width = 183*1.5, height = 240, res = 600)
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure S06
```{r}
# Load individual plots
p.fun <- readRDS(here("Data", "Interim", "Figures", "Shannon_resistance_fungi.rds"))
p.bac <- readRDS(here("Data", "Interim", "Figures", "Shannon_resistance_Bacteria.rds"))

# Combine plots
fig <- plot_grid(p.fun,
                 p.bac,
                 # labels = c("(c)", "(d)"),
                 align = "v",
                 nrow = 2)

fig

# save
png(here("Output", "Figure_S06.png"), units = "mm", 
    width = 183*1.4, height = 120*1.5, res = 600, bg = "white")
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure S07
```{r}
# Load individual plots and legend
p.fun <- readRDS(here("Data", "Interim", "Figures", "Richness_resistance_fungi.rds"))
p.bac <- readRDS(here("Data", "Interim", "Figures", "Richness_resistance_bacteria.rds"))

# Combine plots
fig <- plot_grid(p.fun,
                 p.bac,
                 # labels = c("(c)", "(d)"),
                 align = "v",
                 nrow = 2)

fig

# save
# 89 mm wide (single column) and 183 mm wide (double column)
png(here("Output", "Figure_S07.png"), units = "mm", 
   width = 183*1.4, height = 120*1.5, bg = "white", res = 600)
fig
dev.off()

rm(list = c("p.fun", "p.bac", "fig"))
```

## Figure S08
```{r}
# Load individual plots and legend
p <- readRDS(here("Data", "Interim", "Figures", "PCoA_rri_genotypes_bacteria.rds"))

# save
png(here("Output", "Figure_S08.png"), units = "mm", 
    width = 183, height = 120, res = 600)
p
dev.off()

rm(list = c("p"))
```

## Figure S09
```{r}

# Load individual plot
p <- readRDS(here("Data", "Interim", "Figures", "Herit_daCor_combined.rds"))
p

# save
png(here("Output", "Figure_S09.png"), units = "mm", 
    width = 220, height = 140, res = 600)
p
dev.off()

rm(list = c("p"))
```

## Figure S11
```{r}

# Load individual plot
p <- readRDS(here("Data", "Interim", "Figures", "Prevalence_combined.rds"))
p 

# save
png(here("Output", "Figure_S11.png"), units = "mm", 
    width = 183, height = 100, res = 600)
p
dev.off()


rm(list = c("p"))
```

## Figure S12
```{r}
# Load individual plot
p <- readRDS(here("Data", "Interim", "Figures", "Cor_OTU_RRI.rds"))
p 

# save
png(here("Output", "Figure_S12.png"), units = "mm", 
    width = 183*1.5, height = 120, res = 600)
p
dev.off()

rm(list = c("p"))
```

## Figure S10
```{r}
# load image
load(file = here("Data", "Interim", "Image_Network.RData"))
library(igraph)

# save
png(here("Output", "Figure_S10.png"), units = "mm", width = 183, height = 183, 
    pointsize = 4, res = 800) 
plot(igr_net_comb_red,
     vertex.color = adjustcolor(dtype, alpha.f = 0.75),
     vertex.shape = "circle",
     vertex.frame.color = adjustcolor(dtype.frame, alpha.f = 0.6),
     vertex.frame.width = width.frame,
     vertex.size = 4,
     vertex.label.cex = 0.4,
     vertex.label.color = adjustcolor(dtype.text, alpha.f = 0.75),
     edge.color = adjustcolor("skyblue1", alpha.f = 0.3),
     layout = layout.fruchterman.reingold)
dev.off()
```

```{r}

```
